send-transaction-service/src/transaction_client.rs

```diff
@@ -243,84 +231,67 @@ impl TpuClientNextClient {
-            receiver,
-            update_certificate_receiver,
-            cancel.clone(),
-        );
-        // leaking handle to this task, as it will run until the cancel signal is received
-        runtime_handle.spawn(scheduler.get_stats().report_to_influxdb(
-            "send-transaction-service-TPU-client",
-            METRICS_REPORTING_INTERVAL,
-            cancel.clone(),
-        ));
-        let _handle = runtime_handle.spawn(scheduler.run(config));
+
+        let (sender, client) = ClientBuilder::with_leader_updater(Box::new(leader_updater))
+            .runtime_handle(runtime_handle.clone())
+            .bind_socket(bind_socket)
+            .identity(identity)
+            .leader_send_fanout(leader_forward_count)
+            .max_cache_size(MAX_CONNECTIONS)
+            // The channel size represents 8s worth of transactions at a rate of
+            // 1000 tps, assuming batch size is 64.
+            .sender_capacity(128)
+            .worker_channel_size(64)
+            .max_reconnect_attempts(4)
+            .report({
+                |stats: Arc<SendTransactionStats>, cancel: CancellationToken| async move {
+                    stats
+                        .report_to_influxdb(
+                            "send-transaction-service-TPU-client",
+                            METRICS_REPORTING_INTERVAL,
+                            cancel.clone(),
+                        )
+                        .await
+                }
+            })
+            .cancel_token(cancel)
+            .build::<NonblockingBroadcaster>()
+            .expect("Client should not be misconfigured: invalid bind socket is specified.");
+
         Self {
             runtime_handle,
             sender,
-            update_certificate_sender,
-            #[cfg(any(test, feature = "dev-context-only-utils"))]
-            cancel,
-        }
-    }
-
-    fn create_config(
-        bind_socket: UdpSocket,
-        stake_identity: Option<&Keypair>,
-        leader_forward_count: usize,
-    ) -> ConnectionWorkersSchedulerConfig {
-        ConnectionWorkersSchedulerConfig {
-            bind: BindTarget::Socket(bind_socket),
-            stake_identity: stake_identity.map(StakeIdentity::new),
-            num_connections: MAX_CONNECTIONS,
-            skip_check_transaction_age: true,
-            // experimentally found parameter values
-            worker_channel_size: 64,
-            max_reconnect_attempts: 4,
-            // We open connection to one more leader in advance, which time-wise means ~1.6s
-            leaders_fanout: Fanout {
-                connect: leader_forward_count + 1,
-                send: leader_forward_count,
-            },
+            client,
         }
     }
 
     #[cfg(any(test, feature = "dev-context-only-utils"))]
-    pub fn cancel(&self) {
-        self.cancel.cancel();
+    pub fn shutdown(self) {
+        self.runtime_handle.spawn(async move {
+            self.client.shutdown().await;
+        });
```

Comment:
Is there value in waiting for the shutdown task to finish?
I can see that the `client.shutdown()` seems to be waiting for tasks to finish:

```rust
    pub async fn shutdown(self) {
        self.cancel.cancel();
        drop(self.update_certificate_sender);
        self.tasks.wait().await;
    }
```

Could a wait here remove potential race conditions from the tests?

Reply:
at the moment we don't need it. To achieve this I would need to use `block_on` but the validator might be called in tokio context (local_cluster) and it will lead to panic. In ideal world this SendTransactionService should have been async and in this case `shutdown` would be called directly, but it is not like that at the time.